src/main/haskell/kore/src/Kore/Builtin/Krypto.hs

```diff
@@ -97,3 +111,147 @@ evalKeccak =
+    Builtin.getAttemptedFunction $ do
+        messageHash <- s2i <$> String.expectBuiltinString "" messageHash0
+        v <- Int.expectBuiltinInt "" v0
+        r <- s2i <$> String.expectBuiltinString "" r0
+        s <- s2i <$> String.expectBuiltinString "" s0
+        Builtin.appliedFunction
+            $ String.asExpandedPattern resultSort
+            $ bs2s
+            $ pad 64 0
+            $ signatureToKey messageHash r s v
+eval0 _ _ _ _ = Builtin.wrongArity ecsdaRecover
+
+pad :: Int -> Word8 -> ByteString -> ByteString
+pad n w s = ByteString.append s padding
+  where
+    padding =
+        ByteString.take (n - ByteString.length s)
+      $ ByteString.pack (repeat w)
+
+signatureToKey
+    :: Integer
+    -> Integer
+    -> Integer
+    -> Integer
+    -> ByteString
+signatureToKey messageHash r s v =
+    assert (28 <= v && v <= 34)
+  $ ByteString.drop 1
+  $ encodePoint compressed
+  $ recoverPublicKey recId (r, s) messageHash
+  where
+    recId = v - 27
+    compressed = v >= 31
+
+recoverPublicKey
+    :: Integer
+    -> (Integer, Integer)
+    -> Integer
+    -> Point
+recoverPublicKey recId (r, s) e =
+    assert (recId > 0)
+  $ assert (r > 0)
+  $ assert (s > 0)
+  $ assert (pt_x < p)
+  $ assert (pointMul p256k1 n pt == PointO)
+  $ pointAddTwoMuls
+            p256k1
+            (mulMod n (invMod n r) s)
+            pt
+            (mulMod n (invMod n r) (n - e `mod` n))
+            (ecc_g curveParams)
+  where
+    p256k1@(CurveFP (CurvePrime p curveParams)) = getCurveByName SEC_p256k1
+
+    n = ecc_n curveParams
+
+    i = recId `div` 2
+
+    pt_x = r + i*n
+
+    pt = decompressPt pt_x (recId .&. 1 == 1)
+
+    decompressPt x signBit = Point x (if signBit then y else p - y)
+      where
+        y = sqrtMod p $
+              (powMod p x 3)
+            + (mulMod p (ecc_a curveParams) x)
+            + (ecc_b curveParams)
+
+invMod
+    :: Integer
+    -> Integer
+    -> Integer
+invMod p x = powMod p x (p - 2)
+
+sqrtMod
+    :: Integer
+    -> Integer
+    -> Integer
+sqrtMod p x = powMod p x ((p + 1) `div` 4)
+
+mulMod
+    :: Integer
+    -> Integer
+    -> Integer
+    -> Integer
+mulMod p x y = (x * y) `mod` p
+
+powMod
+    :: Integer
+    -> Integer
+    -> Integer
+    -> Integer
+powMod p _ 0 = 1
+powMod p x a =
+    mulMod p
+    (if even a then 1 else x)
+    (powMod p (mulMod p x x) (a `div` 2))
+
+-- Leading bit signals whether the point is compressed.
+-- Superfluous because we drop it later on.
+-- Kept here for completeness sake, to match
+-- the code in the java backend.
+encodePoint
+    :: Bool
+    -> Point
+    -> ByteString
+encodePoint compressed (Point x y)
+    | compressed =
+        ByteString.cons
+        (if even y then 0x02 else 0x03)
+        (i2bs x)
+    | otherwise = ByteString.concat
+        [ ByteString.pack [0x04]
+        , i2bs x
+        , i2bs y
+        ]
+encodePoint _ _ = error "Should never obtain point-at-infinity here!"
+
+bs2s :: ByteString -> String
```

Comment:
Please give these four utility functions meaningful names. I can infer from context that they are to convert between `ByteString`, `Integer`, and `String`, but what that "conversion" entails is not immediately evident.